### ISSUE TITLE
Support for `ng test` in Angular CLI v6 via angular.json `test` `builder`

### ISSUE THREAD

**Issue by Meligy** [thymikee/jest-preset-angular#162]
In the new Angular CLI `angular.json` file you get this config which backs the `ng test` command:

```
        "test": {
          "builder": "@angular-devkit/build-angular:karma",
          "options": {
            "main": "src/test.ts",
            "polyfills": "src/polyfills.ts",
            "tsConfig": "src/tsconfig.spec.json",
            "karmaConfig": "src/karma.conf.js",
            "styles": [
              "src/styles.css"
            ],
            "scripts": [],
            "assets": [
              "src/favicon.ico",
              "src/assets"
            ]
          }
        },
```

Is it possible create a package similar to `@angular-devkit/build-angular:karma` (not sure about everything it does), and make it available for as a replacement for above?

Thanks for awesome project.

**Comment by Itrulia** [comment]
@ahnpnl this is different. the linked issue is about doing the boilerplate stuff, not overriding the angular cli test command :)

**Comment by ahnpnl** [comment]
I agree. However, @thymikee wanted to keep this preset as simple as possible because it is a preset.
I think about to implement this feature, i'm afraid you have to do it in a separate package :) 

What is your thought @thymikee ?

**Comment by thymikee** [comment]
@ahnpnl is right, I'd rather like something like this to be a separate package which we can link to from docs.

**Comment by Itrulia** [comment]
Yeah I also agree on that, wanted to clarify.

I am currently working on that, hence the issue I opened (encountered while tensting).

**Comment by ahnpnl** [comment]
FYI I found this https://github.com/briebug/jest-schematic?files=1

**Comment by Meligy** [comment]
At the risk of going off topic here, and making it more likely for the maintainer to close this issue, does anyone know the difference between these 2 schematics, and which is better?

- https://github.com/davinkevin/jest
- https://github.com/briebug/jest-schematic

**Comment by ahnpnl** [comment]
I think this question can be on Stackoverflow and it will get more attention and more responses too :)

**Comment by Itrulia** [comment]
@Meligy the first one just installs jest and adds the files but doesn't change the `ng test` command.

**Comment by jogelin** [comment]
It exists a builder : https://github.com/angular-builders/jest 

Maybe it should be merged in this repo ?

**Comment by ahnpnl** [comment]
hi @jogelin , it's not in the scope of this preset. I'd keep it separately. I understand that combining them together into one repo will bring more comfortable to users.

cc @thymikee

**Comment by thymikee** [comment]
We can think about merging at some time, but now it seems like `angular-builders` is not very popular. I'm open to suggestions on how to ease the use and development of this preset :)

**Comment by ahnpnl** [comment]
I see it's a demand for people who want to replace completely the mechanism behind `ng test`. I think it's quite a nice development way for this preset.

**Comment by jogelin** [comment]
@ahnpnl I understand and I ask the thinking of the creator of angular-builders/jest and it's the same point of view

I am just thinking that when a developer wants to use jest in an angular repo today, he shouldn't be lost between `angular-presets` and `schematics to setup jest` and `builder`... 

To me, the best way would be one repo that propose
- default presets to configure jest in angular (like this repo)
- a builder to apply the presets in an angular cli structure (like builder/jes)
- schematics that will remove karma and setup jest

Then with one library, you could setup jest without any configurations...isn't the goal of jest to run the tests without configurations ? :p :p

**Comment by jogelin** [comment]
For people that are using `Nrwl` : https://github.com/nrwl/nx/blob/2659daf8bae892d9a19d84298bc9fce241b55184/packages/builders/src/jest/jest.builder.ts

Everybody is creating a builder for jest, nobody try to work together :p

**Comment by ahnpnl** [comment]
Haha yes 😁

**Comment by meltedspark** [comment]
@jogelin I have to reply here.  
I think the correct way here and in software in general is sticking to SRP principle.  
The purpose of this preset is to indeed provide a **preset** for jest, without any assumptions regarding the workspace. Essentially, this preset has to work with Angular projects that don't use Angular CLI.  
  
On top of it there is [Jest Builder package](https://github.com/meltedspark/angular-builders/tree/master/packages/jest) that takes care of encapsulating this preset and hiding the complexity of setup (but not duplicating the preset work).  
This builder can also contain schematics as schematics is tightly coupled with Angular CLI.  

At the end of the day _you will have a single library that takes care of everything_ and that library would be Jest Builder. The usage would be:
1. `npm i -D @angular-builders/jest`
2. `ng g @angular-builders/jest:setup`

Yet, this preset will remain responsible for a generic Jest configuration for Angular and Jest Builder will remain responsible for hiding the complexity and keeping the workspace clean.

**Comment by Itrulia** [comment]
Hey! I have done something like this here: https://github.com/Itrulia/jest-schematic (as I wrote on 3rd of July I would, I am sorry, it took me a while).


I hope this helps you out!
